### PR TITLE
✨(tests) add end-to-end testing

### PR DESCRIPTION
**Pull request by quitterie-lcs**
# Purpose

End to end testing is necessary to catch logs for the dataset.

# Proposal

- [x] `docker-compose` for cypress
- [x] elementary tests for `cms` and `lms` (login)

### CODE REVIEW

File: e2e/cypress/integration/cms_login_spec.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// CMS Login Test
+
+describe('CMS Login Test', () => {
+    
+    beforeEach(() => {
+        cy.visit('http://localhost:8082/');
```
Comment (jmaupetit) [inline e2e/cypress/integration/cms_login_spec.js:6]: If you use the project network:

```suggestion
        cy.visit('http://edx_cms:8000/');
```

File: e2e/cypress/integration/lms_login_spec.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// LMS Login Test
+
+describe('LMS Login Test', () => {
+  
+    beforeEach(() => {
+      cy.visit('http://localhost:8072/');
```
Comment (jmaupetit) [inline e2e/cypress/integration/lms_login_spec.js:6]:
```suggestion
      cy.visit('http://edx_lms:8000/');
```

File: e2e/cypress/integration/lms_login_spec.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    })
+      
+    it('should fill login form and redirect to dashboard', () => {
+        cy.visit('http://localhost:8072/login');
```
Comment (jmaupetit) [inline e2e/cypress/integration/lms_login_spec.js:10]:
```suggestion
        cy.visit('http://edx_lms:8000/login');
```

File: docker-compose.cypress.yml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+version: "3.5"
+services:
+  cypress:
+    image: "cypress/included:4.0.2"
+    working_dir: /e2e
+    network_mode: host
```
Comment (jmaupetit) [inline docker-compose.cypress.yml:6]: I would recommend to use the project network instead:

```suggestion
```

File: docker-compose.cypress.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+    volumes:
```
Comment (jmaupetit) [inline docker-compose.cypress.yml:7]:
```suggestion
    user: ${DOCKER_UID}:${DOCKER_GID}
    volumes:
```

File: Makefile [hunk 6]
```diff
@@ -43,6 +43,12 @@ clean-db:  ## remove LMS databases
 	$(COMPOSE) rm edx_mongodb edx_mysql edx_redis 
 .PHONY: clean-db
 
+e2e-tests:\
+	run
+e2e-tests: ## run end-to-end tests
```
Comment (jmaupetit) [inline Makefile:48]: I would remove the `run` rule dependency as it's a bit long to execute each time you want to test your app, and `test` is faster to type!

```suggestion
test: ## run end-to-end tests
```

File: Makefile [hunk 7]
```diff
@@ -43,6 +43,12 @@ clean-db:  ## remove LMS databases
...
+	docker-compose -f docker-compose.cypress.yml up
+.PHONY: e2e-tests
```
Comment (jmaupetit) [inline Makefile:50]:
```suggestion
.PHONY: test
```

**Review by jmaupetit — COMMENTED** [review]
Great!

I would add a `bin/compose` and a `bin/cypress` utilities to this project to ease running custom commands:

- `bin/compose` as we've split services in multiple files and need to define the container user,
- `bin/cypress` to at least test browser switch as the default browser is Electron.

File: e2e/cypress/integration/cms_login_spec.js (hunk 1 above)
Comment (quitterie-lcs) [reply e2e/cypress/integration/cms_login_spec.js]: This URL is not found, hence I have to use localhost

File: e2e/cypress/integration/lms_login_spec.js (hunk 3 above)
Comment (quitterie-lcs) [reply e2e/cypress/integration/lms_login_spec.js]: Same as CMS

File: e2e/cypress/integration/cms_login_spec.js (hunk 1 above)
Comment (jmaupetit) [reply e2e/cypress/integration/cms_login_spec.js]: Now it should be if you remove the `network_mode: host` in the docker-compose configuration.

File: e2e/cypress/integration/lms_login_spec.js (hunk 3 above)
Comment (jmaupetit) [reply e2e/cypress/integration/lms_login_spec.js]: ditto

File: e2e/cypress.json [hunk 8]
```diff
@@ -0,0 +1,3 @@
+{
+
+}
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply e2e/cypress.json:3]: Should we remove it?

**Review by jmaupetit — APPROVED** [review]
I only have minor comments :+1:

File: e2e/cypress.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply e2e/cypress.json]: The `--config-file` flag allows to deactivate the use of a json file. Has been updated in the Makefile and `cypress.json` removed

**Pull request merged**